internal/provider/resource_integration.go

```diff
@@ -321,17 +468,24 @@ func generateID(modelName string, apps []juju.Application) string {
 	return id
 }
 
+func modelIntegrationNameAndEndpointsFromID(ID string) (string, string, string, string, string, diag.Diagnostics) {
+	var diags diag.Diagnostics
+	id := strings.Split(ID, ":")
+	if len(id) != 5 {
+		diags.AddError("Malformed ID",
+			fmt.Sprintf("unable to parse model and application name from provided ID: %q", ID))
+		return "", "", "", "", "", diags
+	}
+	return id[0], id[1], id[2], id[3], id[4], diags
+}
+
 // This function can be used to parse the terraform data into usable juju endpoints
 // it also does some sanity checks on inputs and returns user friendly errors
-func parseEndpoints(apps []interface{}) (endpoints []string, offer *string, appNames []string, err error) {
+func parseEndpoints(apps []nestedApplication) (endpoints []string, offer *string, appNames []string, err error) {
```

Comment:
Yeah I noticed that assumption as well. Not sure if that's a bug. Keep in mind that there will be precisely two apps in that slice, so I'm not entirely sure what happens (and what should happen) if we get offer urls from both.



> This entire method should be rewritten for the framework. So many helpers in the frame not being utilized to make it easier.

Not sure if you mean the validation part with this comment. If so, take a look at the `ValidateConfig()` might make you happy. Otherwise, could you please elaborate? I don't think I understand what you mean.

Reply:
Adding the extra ValidateConfig helps. Let's leave it for now, can always adjust later.